errors.go

```diff
@@ -36,3 +27,33 @@ var (
 	// due to floating-point arithmetic.
 	ErrNoProgress = errors.New("linesearch: no change in location after Linesearcher step")
 )
+
+// ErrFuncInit is returned when an initial function value is invalid. The error
+// state may be either +Inf or NaN. ErrFuncInit satisfies the error interface.
+type ErrFuncInit float64
+
+func (err ErrFuncInit) Error() string {
+	switch {
+	case math.IsInf(float64(err), 1):
+		return "optimize: initial function value is infinite"
+	case math.IsNaN(float64(err)):
+		return "optimize: initial function value is NaN"
+	default:
+		panic("optimize: bad ErrFuncInit")
+	}
+}
+
+// ErrGradInit is returned when an initial gradient is invalid. The error state
+// may be either ±Inf or NaN. ErrGradInit satisfies the error interface.
+type ErrGradInit float64
```

Comment:
Would it be overkill to make the type
```go
type ErrGradInit struct {
    v     float64 // Invalid value
    index int     // Index of the value
}
```
(probably with the fields exported)? Seems like the index could be useful information.